src/mito2/src/memtable/merge_tree/tree.rs

```diff
@@ -235,6 +278,132 @@ struct TreeParts {
     /// Index part of the tree. If the region doesn't have a primary key, this field
     /// is `None`.
     index: Option<KeyIndexRef>,
-    /// Data buffer of the tree.
-    data_buffer: DataBuffer,
+    /// Data part of the tree.
+    data: DataParts,
+}
+
+struct ShardIter {
+    metadata: RegionMetadataRef,
+    projection: HashSet<ColumnId>,
+    index_reader: Option<ShardReader>,
+    data_reader: DataReader,
+}
+
+impl Iterator for ShardIter {
+    type Item = Result<Batch>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.data_reader.is_valid() {
+            return None;
+        }
+
+        self.next_batch().transpose()
+    }
+}
+
+impl ShardIter {
+    /// Fetches next batch and advances the iter.
+    fn next_batch(&mut self) -> Result<Option<Batch>> {
+        let Some(index_reader) = &mut self.index_reader else {
+            // No primary key to read.
+            // Safety: `next()` ensures the data reader is valid.
+            let batch = self.data_reader.convert_current_record_batch(
+                &self.metadata,
+                &self.projection,
+                &[],
+            )?;
+            // Advances the data reader.
+            self.data_reader.next()?;
+            return Ok(Some(batch));
+        };
+
+        // Iterate the index reader until we see the same pk index of the data batch.
+        while index_reader.is_valid()
+            && index_reader.current_pk_index() != self.data_reader.current_pk_index()
+        {
+            index_reader.next();
+        }
+        assert!(
+            index_reader.is_valid(),
+            "Data contains pk_index {} not in the index",
+            self.data_reader.current_pk_index()
+        );
+
+        let batch = self.data_reader.convert_current_record_batch(
+            &self.metadata,
+            &self.projection,
+            index_reader.current_key(),
+        )?;
+        // Advances the data reader.
+        self.data_reader.next()?;
+        Ok(Some(batch))
+    }
+}
+
+struct DataReader {
+    current: Option<DataBatch>,
+    iter: data::Iter,
+}
+
+impl DataReader {
+    fn new(mut iter: data::Iter) -> Result<Self> {
+        let current = iter.next().transpose()?;
+
+        Ok(Self { current, iter })
+    }
+
+    fn is_valid(&self) -> bool {
+        self.current.is_some()
+    }
+
+    fn current_pk_index(&self) -> PkIndex {
+        self.current.as_ref().unwrap().pk_index()
+    }
+
+    /// Converts current [RecordBatch] to [Batch].
+    fn convert_current_record_batch(
+        &self,
+        metadata: &RegionMetadataRef,
+        projection: &HashSet<ColumnId>,
+        primary_key: &[u8],
+    ) -> Result<Batch> {
+        let data_batch = self.current.as_ref().unwrap();
+        let offset = data_batch.range().start;
+        let length = data_batch.range().end - offset;
```

Comment:
```suggestion
        let length = data_batch.range().len();
```

Reply:
Addressed in https://github.com/evenyag/greptimedb/pull/14/commits/252f7da1de2703ae1bb2ab5318b2557e3effc46a